consensus/polybft/proposer_calculator.go

```diff
@@ -0,0 +1,499 @@
+	// total voting power gives the maximum allowed distance between validator
+	// priorities.
+	priorityWindowSizeFactor = 2
+)
+
+var (
+	// errInvalidTotalVotingPower is returned if the total voting power is zero
+	errInvalidTotalVotingPower = errors.New(
+		"invalid voting power configuration provided: total voting power must be greater than 0")
+)
+
+// Holds ValidatorMetadata together with priority
+type PrioritizedValidator struct {
+	Metadata         *ValidatorMetadata
+	ProposerPriority int64
+}
+
+// ProposerSnapshot represents snapshot of one proposer calculation
+type ProposerSnapshot struct {
+	Height     uint64
+	Round      uint64
+	Proposer   *PrioritizedValidator
+	Validators []*PrioritizedValidator
+}
+
+// NewProposerSnapshotFromState create ProposerSnapshot from state if possible or from genesis block
+func NewProposerSnapshotFromState(config *runtimeConfig, logger hclog.Logger) (*ProposerSnapshot, error) {
+	snapshot, err := config.State.getProposerSnapshot()
+	if err != nil {
+		return nil, err
+	}
+
+	if snapshot == nil {
+		// pick validator set from genesis block if snapshot is not saved in db
+		genesisValidatorsSet, err := config.polybftBackend.GetValidators(0, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		snapshot = NewProposerSnapshot(1, genesisValidatorsSet)
+	}
+
+	return snapshot, nil
+}
+
+// NewProposerSnapshot creates ProposerSnapshot with height and validators with all priorities set to zero
+func NewProposerSnapshot(height uint64, validators []*ValidatorMetadata) *ProposerSnapshot {
+	validatorsSnap := make([]*PrioritizedValidator, len(validators))
+
+	for i, x := range validators {
+		validatorsSnap[i] = &PrioritizedValidator{Metadata: x, ProposerPriority: int64(0)}
+	}
+
+	return &ProposerSnapshot{
+		Round:      0,
+		Proposer:   nil,
+		Height:     height,
+		Validators: validatorsSnap,
+	}
+}
+
+// Gets total voting power from all the validators
+func (pcs ProposerSnapshot) GetTotalVotingPower() int64 {
+	totalVotingPower := int64(0)
+
+	for _, v := range pcs.Validators {
+		totalVotingPower = safeAddClip(totalVotingPower, int64(v.Metadata.VotingPower))
+	}
+
+	return totalVotingPower
+}
+
+// Returns copy of current ProposerSnapshot object
+func (pcs *ProposerSnapshot) Copy() *ProposerSnapshot {
+	var proposer *PrioritizedValidator
+
+	valCopy := make([]*PrioritizedValidator, len(pcs.Validators))
+
+	for i, val := range pcs.Validators {
+		valCopy[i] = &PrioritizedValidator{Metadata: val.Metadata.Copy(), ProposerPriority: val.ProposerPriority}
+
+		if pcs.Proposer != nil && pcs.Proposer.Metadata.Address == val.Metadata.Address {
+			proposer = valCopy[i]
+		}
+	}
+
+	return &ProposerSnapshot{
+		Validators: valCopy,
+		Height:     pcs.Height,
+		Round:      pcs.Round,
+		Proposer:   proposer,
+	}
+}
+
+// ProposerCalculator interface - proposer calculator algorithm should implement this interface
+type ProposerCalculator interface {
```

Comment:
Why do we need this interface here? For testing?

Reply:
The idea was simmilar as for `ValidatorSet`, to have an interface if we want to change algorithm. Since we are moving `CalcProposer` to `ProposerSnapshot`, I'll remove the interface

Reply:
Fixed https://github.com/0xPolygon/polygon-edge/pull/1018/commits/48f6b5d8b23ffd0eb1cfc40fe6aa9ccfd9188845